feat(regex): add support for regex matchers for dynamic params

thetutlage opened the pull request:
I have added the support to define matchers to match the dynamic parameters during the `match` operation. ( addresses #9  )

## What are matchers?
Matchers are key/value pairs to define regular expression for matching the params value. The `key` is the param name and `value` is the regex to test it against.

```js
const matchers = {
  bar: /[a-z]/
}

parse('/foo/:bar', matchers)
```

## How it works?

1. Matchers can only be defined for dynamic params (including optionals).
2. A matcher cannot be defined for a wildcard.
3. During the `parse` operation, the matcher for the current value will be attached to it's `node`.
4. During `match` if `matcher` property exists, it will be tested against that.

## PR

- [x] I have added the required tests.
- [ ] Update REAMDE (Not yet, if you are interested in the change, I can update it).
- [ ] Update Benchmarks (Not yet, if you are interested in the change, I can update it).

thetutlage committed:
feat(regex): add support for regex matchers for dynamic params

lukeed reviewed (commented):
Hey, thanks~! 

Before I take an in-depth look at this, do you mind posting before & after benchmarks? 
Thank you!

thetutlage commented:
Sure I’ll.

One more question. Do you expect people to primarly use matchit for creating HTTP routes? 

If so, then have you thought about associating HTTP method within the generated tokens list?

lukeed commented:
Thanks!

No, this is primarily _and only_ a string pattern parser – the strings just happen to look like server routes in the examples. I use this on the browser in a few places, for example.

[Trouter](https://github.com/lukeed/trouter) may be more what you're after, which builds atop `matchit` and serves as the core router of [Polka](https://github.com/lukeed/polka).

lukeed commented:
Very cool 🎉 looking forward to seeing that.

As you can imagine, I'm very hesitant to make changes to this library simply because it, alone, is the reason why Polka, Rayo, and now Adonis are so quick

thetutlage commented:
Sure!

Here are the benchmarks and covers following use cases.

1. Old unchanged code
2. Changed code but not using any matchers
3. Changed code with matchers.

I personally don't see any overhead, since results are more of less the same.

<details><summary> Old unchanged code </summary>
<pre><code>
matchit.parse x 488,920 ops/sec ±2.68% (94 runs sampled)
path-to-regexp x 72,742 ops/sec ±1.55% (90 runs sampled)
path-to-regexp.parse x 375,198 ops/sec ±1.63% (88 runs sampled)
Fastest is matchit.parse


matchit.match (index) x 72,723,890 ops/sec ±2.53% (87 runs sampled)
path-to-regexp.exec (index) x 5,080,630 ops/sec ±9.43% (73 runs sampled)
path-to-regexp.tokens (index) x 60,369 ops/sec ±7.49% (72 runs sampled)
Fastest is matchit.match (index)


matchit.match (param) x 2,231,677 ops/sec ±3.19% (86 runs sampled)
path-to-regexp.exec (param) x 3,679,983 ops/sec ±5.40% (77 runs sampled)
path-to-regexp.tokens (param) x 67,940 ops/sec ±8.18% (70 runs sampled)
Fastest is path-to-regexp.exec (param)


matchit.match (optional) x 1,275,827 ops/sec ±13.56% (54 runs sampled)
path-to-regexp.exec (optional) x 6,169,748 ops/sec ±7.42% (75 runs sampled)
path-to-regexp.tokens (optional) x 71,722 ops/sec ±8.32% (74 runs sampled)
Fastest is path-to-regexp.exec (optional)


matchit.match (wildcard) x 2,651,735 ops/sec ±4.64% (80 runs sampled)
path-to-regexp.exec (wildcard) x 7,012,970 ops/sec ±5.16% (83 runs sampled)
path-to-regexp.tokens (wildcard) x 87,108 ops/sec ±1.09% (88 runs sampled)
Fastest is path-to-regexp.exec (wildcard)


matchit.exec (params) x 1,467,742 ops/sec ±1.36% (92 runs sampled)
path-to-regexp.exec (params) x 86,385 ops/sec ±0.99% (90 runs sampled)
Fastest is matchit.exec (params)
</code></pre>
</details>

---

<details><summary> Changed code but not using any matchers </summary>
<pre><code>
matchit.parse x 535,861 ops/sec ±2.81% (92 runs sampled)
path-to-regexp x 73,531 ops/sec ±0.78% (94 runs sampled)
path-to-regexp.parse x 378,655 ops/sec ±1.38% (88 runs sampled)
Fastest is matchit.parse

matchit.match (index) x 76,978,943 ops/sec ±0.61% (94 runs sampled)
path-to-regexp.exec (index) x 6,332,324 ops/sec ±1.72% (91 runs sampled)
path-to-regexp.tokens (index) x 85,254 ops/sec ±0.86% (92 runs sampled)
Fastest is matchit.match (index)

matchit.match (param) x 2,416,169 ops/sec ±0.35% (93 runs sampled)
path-to-regexp.exec (param) x 4,257,386 ops/sec ±1.53% (91 runs sampled)
path-to-regexp.tokens (param) x 87,245 ops/sec ±0.24% (98 runs sampled)
Fastest is path-to-regexp.exec (param)

matchit.match (optional) x 3,091,059 ops/sec ±1.48% (91 runs sampled)
path-to-regexp.exec (optional) x 7,580,495 ops/sec ±1.03% (88 runs sampled)
path-to-regexp.tokens (optional) x 65,185 ops/sec ±9.33% (70 runs sampled)
Fastest is path-to-regexp.exec (optional)

matchit.match (wildcard) x 3,020,901 ops/sec ±1.66% (89 runs sampled)
path-to-regexp.exec (wildcard) x 7,434,808 ops/sec ±2.14% (88 runs sampled)
path-to-regexp.tokens (wildcard) x 86,595 ops/sec ±0.73% (91 runs sampled)
Fastest is path-to-regexp.exec (wildcard)

matchit.exec (params) x 1,536,656 ops/sec ±0.29% (96 runs sampled)
path-to-regexp.exec (params) x 86,005 ops/sec ±1.59% (92 runs sampled)
Fastest is matchit.exec (params)
</code></pre>
</details>

---

<details><summary> Changed code with matchers </summary>
<pre><code>
matchit.parse x 697,614 ops/sec ±4.15% (86 runs sampled)
path-to-regexp x 73,280 ops/sec ±0.97% (92 runs sampled)
path-to-regexp.parse x 389,722 ops/sec ±1.98% (88 runs sampled)
Fastest is matchit.parse


matchit.match (index) x 74,606,161 ops/sec ±2.64% (87 runs sampled)
path-to-regexp.exec (index) x 6,615,060 ops/sec ±1.08% (94 runs sampled)
path-to-regexp.tokens (index) x 87,720 ops/sec ±1.18% (95 runs sampled)
Fastest is matchit.match (index)


matchit.match (param) x 2,420,655 ops/sec ±1.18% (93 runs sampled)
path-to-regexp.exec (param) x 4,334,384 ops/sec ±2.19% (91 runs sampled)
path-to-regexp.tokens (param) x 86,348 ops/sec ±1.36% (88 runs sampled)
Fastest is path-to-regexp.exec (param)

matchit.match (optional) x 3,115,418 ops/sec ±1.38% (94 runs sampled)
path-to-regexp.exec (optional) x 8,057,054 ops/sec ±1.49% (90 runs sampled)
path-to-regexp.tokens (optional) x 86,192 ops/sec ±1.52% (91 runs sampled)
Fastest is path-to-regexp.exec (optional)


matchit.match (wildcard) x 3,099,154 ops/sec ±0.53% (92 runs sampled)
path-to-regexp.exec (wildcard) x 8,028,428 ops/sec ±0.54% (92 runs sampled)
path-to-regexp.tokens (wildcard) x 86,949 ops/sec ±0.76% (91 runs sampled)
Fastest is path-to-regexp.exec (wildcard)


matchit.exec (params) x 1,393,131 ops/sec ±0.87% (90 runs sampled)
path-to-regexp.exec (params) x 87,581 ops/sec ±0.66% (94 runs sampled)
Fastest is matchit.exec (params)
</code></pre>
</details>

thetutlage commented:
@lukeed Got a chance to look into it?

lukeed commented:
This is at the top of my list for OSS this week 💯 I will get to it with sufficient testing and benchmarking. I was traveling all weekend, otherwise this'd be answered by now :)

Thank you, sorry for the delay(s)! 🙇

lukeed commented:
Hey,

So the `parse` slowed down considerably, but the `match` & `exec` speeds remained the same. I'll look into speeding up `parse` right now

```
Node 8
---

(before)
matchit.parse x 1,100,818 ops/sec ±0.78% (95 runs sampled)
path-to-regexp x 75,116 ops/sec ±0.82% (94 runs sampled)
path-to-regexp.parse x 385,112 ops/sec ±0.59% (96 runs sampled)

(after)
matchit.parse x 531,861 ops/sec ±1.19% (88 runs sampled)
path-to-regexp x 76,251 ops/sec ±0.77% (89 runs sampled)
path-to-regexp.parse x 394,572 ops/sec ±0.93% (95 runs sampled)


Node 10.12
---

(before)
matchit.parse x 1,421,908 ops/sec ±5.78% (90 runs sampled)
path-to-regexp x 81,402 ops/sec ±0.89% (94 runs sampled)
path-to-regexp.parse x 415,936 ops/sec ±1.46% (91 runs sampled)

(after)
matchit.parse x 580,865 ops/sec ±2.46% (92 runs sampled)
path-to-regexp x 80,181 ops/sec ±0.72% (90 runs sampled)
path-to-regexp.parse x 422,954 ops/sec ±1.29% (89 runs sampled)
```

lukeed committed:
share `val` placeholder

lukeed committed:
inline `getMatcher` fn

lukeed committed:
fix: only look for “matcher” w/ params

lukeed committed:
fix: only expect “matcher” key when P|OTYPE

lukeed commented:
Okay, with those changes here are the latest `parse` numbers:

```
Node 8
---

matchit.parse x 766,155 ops/sec ±1.77% (86 runs sampled)
path-to-regexp x 75,994 ops/sec ±0.82% (92 runs sampled)
path-to-regexp.parse x 389,971 ops/sec ±0.37% (89 runs sampled)


Node 10
---
matchit.parse x 1,061,875 ops/sec ±3.54% (93 runs sampled)
path-to-regexp x 81,485 ops/sec ±0.78% (93 runs sampled)
path-to-regexp.parse x 403,606 ops/sec ±0.64% (91 runs sampled)
```

I also fixed the extra `getMatcher` call that was running alongside `STYPE` definitions. Even by your own definition, there should only be a `matcher` for parameters & nothing else. Since there was only 1 call to `getMatcher`, I inlined the function – it was just a lookup anyway & there was no reason to continually assert that `key == RegExp` so I moved it to the beginning.

TBH, not sure how I feel about having custom errors in here. I could go either way on it, but I generally don't like having type-checking in runtime code. In my (not very good) opinion, that's what testing/development is for.

thetutlage commented:
I believe, the reason for type checking is to find the correct route and not to validate the params. For example:

```js
parse('/foo/:name', {
  name: /[a-z]/
})

parse('/foo/:id', {
  id: /[0-9]/
})
```

Now when I match for `/foo/1`. I want to match the 2nd expression and not 1st.

lukeed commented:
Right, but the type checking I'm talking about is the one that makes sure both `id` and `name` are RegExp

thetutlage commented:
Okay yeah makes sense. I believe checking for `Regex` within the parse function makes more sense vs doing it in match.
